lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,260 @@
+from streamlit.connections.util import load_from_snowsql_config_file, running_in_sis
+from streamlit.errors import StreamlitAPIException
+from streamlit.runtime.caching import cache_data
+
+if TYPE_CHECKING:
+    from snowflake.connector import (  # type: ignore # isort: skip
+        SnowflakeConnection as InternalSnowflakeConnection,
+    )
+    from snowflake.connector.cursor import SnowflakeCursor  # type: ignore
+    from snowflake.snowpark.session import Session  # type: ignore
+
+
+_REQUIRED_CONNECTION_PARAMS = {"account"}
+
+
+def _validate_connection_params(params: Dict[str, Any]) -> None:
+    for p in _REQUIRED_CONNECTION_PARAMS:
+        if p not in params:
+            raise StreamlitAPIException(f"Missing Snowflake connection param: {p}")
+
+
+class SnowflakeConnection(BaseConnection["InternalSnowflakeConnection"]):
+    """A connection to Snowflake using the Snowflake Python Connector. Initialize using
+    ``st.connection("<name>", type="snowflake")``.
+
+    SnowflakeConnection supports direct SQL querying using ``.query("...")``, access to
+    the underlying Snowflake Python Connector object with ``.raw_connection``, and other
+    convenience functions. See the methods below for more information.
+    SnowflakeConnections should always be created using ``st.connection()``, **not**
+    initialized directly.
+    """
+
+    def _connect(self, **kwargs) -> "InternalSnowflakeConnection":
+        import snowflake.connector  # type: ignore
+        from snowflake.snowpark.context import get_active_session  # type: ignore
+
+        # If we're running in SiS, just call get_active_session() and retrieve the
+        # lower-level connection from it.
+        if running_in_sis():
+            session = get_active_session()
+
+            if hasattr(session, "connection"):
+                return session.connection
+            return session._conn._conn
+
+        # Otherwise, attempt to create a new connection from whatever credentials we
+        # have available.
+
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+
+            # the kwargs passed to `st.connection`.
+            conn_kwargs = {**st_secrets, **kwargs}
+            _validate_connection_params(conn_kwargs)
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            # TODO(vdonato): Double-check that we want to map the "snowflake" connection
+            # name in a call to `st.connection` to the "default" connection in the
+            # Snowflake config file. This is aligned with the default behavior for the
+            # underlying python connector and seems more sane than having a "snowflake"
+            # connection in a snowflake-specific config file.
+            connection_name = (
+                "default"
```

Comment:
You can also retrieve the default connection name from the CONFIG_MANAGER like this:
```suggestion
                snowflake.connector.connection.CONFIG_MANAGER["default_connection_name"]
```

Reply:
I was playing with this, I think we should check for CONFIG_MANAGER["default_connection_name"] if available and then fall back to "default" as the value otherwise.

It seems to work nicely, I was able to do `snow connection add` with snowcli, create a connection named "default", and then `st.connection("snowflake")` just worked. Same for connections of other types. This is really slick!!!

Reply:
also since the default connection name value comes from `SNOWFLAKE_DEFAULT_CONNECTION_NAME`, I think it means a developer could just add that to their secrets.toml to point to whatever connection they have defined, and just use `st.connection("snowflake")`. This might be a more common pattern for SiS apps being developed locally.